Address comments remaining from #25353

ariard opened the pull request:
This PR should address the remaining comments from #25353.

ariard committed:
Address comments remaining from #25353

sdaftuar commented:
> A opt-out child of
a replaceable parent should be from now on seen as non-replaceable.

I think there's some confusion about why this RPC was implemented this way in the first place.  The purpose was to allow a node operator to be able to determine whether a given transaction might be replaceable due to BIP 125 RBF semantics. Since replacing a parent causes all descendants to be removed from the mempool, this logic checks all the unconfirmed ancestors of a transaction to see if they are replaceable.

I understand the confusion this has created for developers who may have understood the BIP to mean that it should be sufficient to conflict with such a non-signaling child transaction and be able to achieve RBF semantics. However, I think that is an oversight in documentation and maybe our release notes for what is intended with this logic, and not a mistake in the logic itself.  I think as long as we support BIP 125 RBF, this RPC is doing what is intended (and hopefully we will eventually drop BIP 125 altogether and we can eliminate this sort of confusion!).

jonatack reviewed (commented):
Some feedback

MarcoFalke commented:
NACK. A bool has two states, and the confusion around this is based on a misinterpretation of at least three states (tx can be replaced "by itself", tx can be replaced "via parent", tx can't be bip125 replaced). Toggling the bool doesn't help anything in fixing the underlying misunderstanding. If you really want to fix this you may offer the consumer a way to determine the state. I think it should be sufficient to just return the two boolean values `SignalsOptInRBF` and `IsRBFOptIn`, alternatively you can return a new field (something like https://github.com/bitcoin/bitcoin/pull/16490).

glozow commented:
@ariard Perhaps consider limiting the scope of this PR to "Address comments remaining from #25353" ?

ariard commented:
Updated at 1056bbdfc with comments addressed and dropped the fix `getmempoolentry`.

----------

So from reviewing back #22698, it wasn't clear to me what was the state of the thinking on fixing the discrepancy between the BIP 125 documentation and our replacement logic. Effectively, if a non-signaling child transaction marked as "bip125-replaceable" should be interpreted as "tx can be replaced via parent', I agree it can be an acceptable-though-non-perfect result. It might still confuses developers attempting to replace the non-signaling child via an assumed "inherited signaling" and I believe it would be good to update or refresh BIP125 on that regard (though I don't plan to do so). It should be noted "inherited signaling" is implemented by other[ implementatio](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-May/018926.html)n so I think any update to BIP125 might require coordination beyond the scope of the project. As suggested, we could enrich `getmempoolentry` with a new field on that subtlety but I don't have an interest there.

MarcoFalke commented on test/functional/feature_rbf.py:
```diff
@@ -702,17 +702,16 @@ def test_replacement_relay_fee(self):
             from_node=self.nodes[0],
             utxo_to_spend=confirmed_utxo,
-            sequence=SEQUENCE_FINAL,
+            sequence=BIP125_SEQUENCE_NUMBER + 1,
```
Is there an explanation why the arithmetic is easier to understand than a named constant?

glozow replied:
Not extremely opinionated about this, but I interpret SEQUENCE_FINAL as primarily for disabling locktime, and the fact that it doesn't signal rbf is a side effect. bip125signal + 1 literally means no signal. But fine with whatever is conventionally used / easiest to understand.

MarcoFalke replied:
Ah right. I wonder why `BIP125_SEQUENCE_NUMBER` isn't called `MAX_BIP125_RBF_SEQUENCE`?

MarcoFalke commented:
cr ACK 1056bbdfcd40b6bc4e16844c6da5cf666e87b1bc

glozow commented:
ACK 1056bbdfcd40b6bc4e16844c6da5cf666e87b1bc

Thanks for splitting the PRs. Confirmed this resolves:
- https://github.com/bitcoin/bitcoin/pull/25353#discussion_r915676651
- https://github.com/bitcoin/bitcoin/pull/25353#discussion_r916128167
- https://github.com/bitcoin/bitcoin/pull/25353#issuecomment-1179122432
- https://github.com/bitcoin/bitcoin/pull/25353#discussion_r915679115
- https://github.com/bitcoin/bitcoin/pull/25353#discussion_r915692270
- https://github.com/bitcoin/bitcoin/pull/25353#discussion_r915704128

CI error in wallet_encryption.py is unrelated, since this only touches feature_rbf.py and docs.

w0xlt reviewed (approved):
cr ACK https://github.com/bitcoin/bitcoin/pull/25575/commits/1056bbdfcd40b6bc4e16844c6da5cf666e87b1bc
